flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java

```diff
@@ -0,0 +1,317 @@
+				name(), checkpointId, pendingTransaction);
+
+			// If this fails, there is actually a data loss
+			commit(pendingTransaction.transaction);
+
+			LOG.debug("{} - committed checkpoint transaction {}", name(), pendingTransaction);
+		}
+
+		pendingCommitTransactions.clear();
+		for (TransactionAndCheckpoint<TXN> remainingTransaction : remainingTransactions) {
+			pendingCommitTransactions.add(remainingTransaction);
+		}
+	}
+
+	@Override
+	public final void snapshotState(FunctionSnapshotContext context) throws Exception {
+		// this is like the pre-commit of a 2-phase-commit transaction
+		// we are ready to commit and remember the transaction
+
+		checkState(currentTransaction != null, "bug: no transaction object when performing state snapshot");
+
+		long checkpointId = context.getCheckpointId();
+		LOG.debug("{} - checkpoint {} triggered, flushing transaction '{}'", name(), context.getCheckpointId(), currentTransaction);
+
+		preCommit(currentTransaction);
+		pendingCommitTransactions.add(new TransactionAndCheckpoint<>(currentTransaction, checkpointId));
+		LOG.debug("{} - stored pending transactions {}", name(), pendingCommitTransactions);
+
+		currentTransaction = beginTransaction();
+		LOG.debug("{} - started new transaction '{}'", name(), currentTransaction);
+
+		pendingCommitTransactionsState.clear();
+		for (TransactionAndCheckpoint<TXN> pendingCommitTransaction : pendingCommitTransactions) {
+			pendingCommitTransactionsState.add(pendingCommitTransaction);
+		}
+
+		pendingTransactionsState.clear();
+		// in case of failure we might not be able to abort currentTransaction. Let's store it into the state
+		// so it can be aborted after a restart/crash
+		pendingTransactionsState.add(currentTransaction);
+	}
+
+	@Override
+	public final void initializeState(FunctionInitializationContext context) throws Exception {
+		// when we are restoring state with pendingCommitTransactions, we don't really know whether the
+		// transactions were already committed, or whether there was a failure between
+		// completing the checkpoint on the master, and notifying the writer here.
+
+		// (the common case is actually that is was already committed, the window
+		// between the commit on the master and the notification here is very small)
+
+		// it is possible to not have any transactions at all if there was a failure before
+		// the first completed checkpoint, or in case of a scale-out event, where some of the
+		// new task do not have and transactions assigned to check)
+
+		// we can have more than one transaction to check in case of a scale-in event, or
+		// for the reasons discussed in the 'notifyCheckpointComplete()' method.
+
+		pendingCommitTransactionsState = context.getOperatorStateStore().getSerializableListState("pendingCommitTransactions");
+		pendingTransactionsState = context.getOperatorStateStore().getSerializableListState("pendingTransactions");
```

Comment:
`getSerializableListState` is deprecated and discouraged usage.
I would recommend that implementations may also pass in either the `TypeInformation` or their own `TypeSerializer` for the transaction state holder.